src/bin/ipa_bench/attribute.rs

```diff
@@ -0,0 +1,401 @@
+use crate::gen_events::{Event, EventBase, MatchKey, SourceEvent, TriggerEvent};
+use log::info;
+use raw_ipa::helpers::models::SecretSharable;
+use serde::Serialize;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+use std::time::Instant;
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize))]
+struct Report {
+    breakdown_key: String,
+    matchkeys: MatchKey,
+    reach: u32,
+    impressions: u32,
+    conversions: u32,
+    total_value: u32,
+}
+
+impl Report {
+    fn new(key: &str) -> Self {
+        Report {
+            breakdown_key: String::from(key),
+            matchkeys: Vec::new(),
+            reach: 0,
+            impressions: 0,
+            conversions: 0,
+            total_value: 0,
+        }
+    }
+}
+
+pub fn generate_report(
+    input: &mut Box<dyn io::Read>,
+    attribution_window: u32,
+    model: &str,
+    out: &mut Box<dyn io::Write>,
+) {
+    let reports = gen_report_impl(input, attribution_window, model);
+
+    // Output as CSV. We can add --output-format option in the future
+    writeln!(out, "breakdown_key,reach,impressions,conversions,values").unwrap();
+    for r in &reports {
+        writeln!(
+            out,
+            "{},{},{},{},{}",
+            r.breakdown_key, r.reach, r.impressions, r.conversions, r.total_value
+        )
+        .unwrap();
+    }
+}
+
+fn gen_report_impl(
+    input: &mut Box<dyn io::Read>,
+    attribution_window: u32,
+    model: &str,
+) -> Vec<Report> {
+    let buf = io::BufReader::new(input);
+
+    let mut events: Vec<Event> = Vec::new();
+
+    // 1. Sort matchkeys in each event. Everything in memory
+    let start = Instant::now();
+
+    for line in buf.lines() {
+        let l = line.unwrap();
```

Comment:
```suggestion
        let l = line.unwrap().trim();
```

that way, you can remove the 2 other uses of `trim` below